NWL/NWL.py

```diff
@@ -0,0 +1,451 @@
+def plot(NWL_mat, phi_bins, theta_bins, num_levels):
+    """Generates contour plot of NWL.
+
+    Arguments:
+        NWL_mat (array of array of float): NWL solutions at centroids of
+            (phi, theta) bins (MW).
+        phi_bins (array of float): centroids of toroidal angle bins (rad).
+        theta_bins (array of float): centroids of poloidal angle bins (rad).
+        num_levels (int): number of contour regions.
+    """
+    import matplotlib.pyplot as plt
+    
+    phi_bins = np.rad2deg(phi_bins)
+    theta_bins = np.rad2deg(theta_bins)
+
+    levels = np.linspace(np.min(NWL_mat), np.max(NWL_mat), num = num_levels)
+    fig, ax = plt.subplots()
+    CF = ax.contourf(phi_bins, theta_bins, NWL_mat, levels = levels)
+    cbar = plt.colorbar(CF)
+    cbar.ax.set_ylabel('NWL (MW)')
+    plt.xlabel('Toroidal Angle (degrees)')
+    plt.ylabel('Poloidal Angle (degrees)')
+    fig.savefig('NWL.png')
+
+
+def NWL_plot(
+    source_file, ss_file, plas_eq, tor_ext, pol_ext, wall_s, num_phi = 101, num_theta = 101, num_levels = 10
+    ):
+    """Computes and plots NWL.
+
+    Arguments:
+        source_file (str): path to OpenMC surface source file.
+        ss_file (str): source strength input file.
+        plas_eq (str): path to plasma equilibrium NetCDF file.
+        tor_ext (float): toroidal extent of model (deg).
+        pol_ext (float): poloidal extent of model (deg).
+        wall_s (float): closed flux surface label extrapolation at wall.
+        num_phi (int): number of toroidal angle bins (defaults to 101).
+        num_theta (int): number of poloidal angle bins (defaults to 101).
+        num_levels (int): number of contour regions (defaults to 10).
+    """
+    import read_vmec
+    
+    tor_ext = np.deg2rad(tor_ext)
+    pol_ext = np.deg2rad(pol_ext)
+    
+    coords = extract_coords(source_file)
+    
+    # Load plasma equilibrium data
+    vmec = read_vmec.vmec_data(plas_eq)
+
+    phi_bins = np.linspace(0.0, tor_ext, num = num_phi)
+    theta_bins = np.linspace(-pol_ext/2, pol_ext/2, num = num_theta)
+    
+    # Initialize count matrix
+    count_mat = np.zeros((num_phi, num_theta))
+
+    for pt in coords:
+        # Extract Cartesian coordinates and format as array
+        pt = [pt['x'], pt['y'], pt['z']]
+        pt = np.array(pt)
+        
+        phi, theta = flux_coords(vmec, wall_s, pt)
+        
+        # Shift angles to fit in bins
+        if theta > pol_ext/2:
+            theta = theta - pol_ext
+
+        for i, phi_bin in enumerate(phi_bins):
+            # Conditionally contribute to bin crossing count if crossing within
+            # bin
+            if np.abs(phi - phi_bin) <= tor_ext/(num_phi - 1)/2:
+                for j, theta_bin in enumerate(theta_bins):
+                    # Conditionally contribute to bin crossing count if
+                    # crossing within bin
+                    if np.abs(theta - theta_bin) <= pol_ext/(num_theta - 1)/2:
+                        count_mat[i,j] += 1
```

Comment:
Can we use Numpy's `histogram2d` for this? https://numpy.org/doc/stable/reference/generated/numpy.histogram2d.html